### PR TITLE
Add edit in staged payment

### LINKED ISSUE #3795: Expenditures: Staged payment - add edit option to locked state
[Provided by Adrian via Contractors project management platform]

## Description
User has to be able to edit staged payment.

## Acceptance Criteria

- the ability to change the recipient
- the ability to change the value of the staged payment and the token selected
- the ability to change the name of a milestone
- the ability to change the percent allocation of a milestone
- the ability to add and remove milestones

## Design

- [Figma Link - Staged payments edit state](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=6182%3A337650)

Screenshot 2022-09-01 123841

### PR DESCRIPTION
**Pull request by agnieszkajarosikloj**
## Description

This PR adds edit in staged payment. 
Edit in other payment types will be added in separate PRs.

**Acceptance Criteria** 

- the ability to change the recipient
- the ability to change the value of the staged payment and the token selected
- the ability to change the name of a milestone
- the ability to change the percent allocation of a milestone
- the ability to add and remove milestones

Screenshot 2022-10-05 at 12 30 16

Resolves  #3795

### CODE REVIEW

**Review by willm30 — CHANGES_REQUESTED** [review]
Nice stuff. Got a few notes. 

I think the heading for milestone could read: "Change milestone: 1" because all the other headings have "Change" in them. 

Also, do we need it to read "Change to", instead of just "To"? We could avoid repetition of "change" since it's included in the heading. (Would apply to them all, in all edit modals)

editstaged

I'm noticing this behaviour when scrolling within the milestone input. This can be fixed if you increase the height of the input slightly. 
milestone

The space between the token symbol and the text is 10px in figma. 
spac

The selected recipient for To is missing the "@" before the name: 
missingat

And the cursor when hovering over the edit icon should be `cursor: pointer` whereas it's currently default.

**Comment by agnieszkajarosikloj** [comment]
@willm30 I added fixes:

- changed headers and "To" label, 
- added fixes to styles,
- changed label in UserPickerWithSearch component -  added "@" before the name.

Can you re-review please?

**Review by chinins — CHANGES_REQUESTED** [review]
Mostly all good, but there is one part that needs refactoring in my opinion. Please, let me know once done and I'll re-review straight away.

**Comment by agnieszkajarosikloj** [comment]
@chinins I have added changes. Can you have a look please?

**Review by chinins — APPROVED** [review]
Thank you - that's so much better! Just, please, wrap the 2 variables in useMemo (as per my direct comments).

File: src/modules/pages/components/ExpenditurePage/utils.ts [hunk 1]
```diff
@@ -312,6 +312,66 @@ export const updateValues = (values, confirmedValues) => {
...
+    // milestones without any changes
+    const sameMilestones = values.staged.milestones.filter((milestone) => {
+      if (
+        !confirmedValues.staged.milestones ||
+        !confirmedValues.staged.milestones?.length
+      ) {
+        return true;
+      }
+      return confirmedValues.staged.milestones?.every(
+        (confirmedMielstone) => confirmedMielstone.id !== milestone.id,
```
Comment (jakubcolony) [inline src/modules/pages/components/ExpenditurePage/utils.ts:332]: Typo

File: src/modules/pages/components/ExpenditurePage/utils.ts [hunk 2]
```diff
@@ -312,6 +312,66 @@ export const updateValues = (values, confirmedValues) => {
...
+    // 'confirmedMilestones' is an array with updated milestones. It is composed of:
+    // - milestones that haven't changed
+    // - updated milestones
+    // - newly created milestones.
+    // And we need to filter out deleted items (with 'removed' property set to true)
+    const confirmedMilestones = [
+      ...sameMilestones,
+      ...changedMilestones?.map((milestone) => {
+        const newValue = confirmedValues.staged.milestones?.find(
+          (confirmedMielstone) => confirmedMielstone.id === milestone.id,
```
Comment (jakubcolony) [inline src/modules/pages/components/ExpenditurePage/utils.ts:349]: Typo

**Review by jakubcolony — CHANGES_REQUESTED** [review]
Just spotted a typo in the screenshot Will posted earlier so I highlighted all of its occurrences

File: src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedStaged/ChangedMilestone/ChangedMilestone.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import { defineMessages, FormattedMessage } from 'react-intl';
+
+import Icon from '~core/Icon';
+import { Milestone } from '~dashboard/ExpenditurePage/Staged/types';
+import { Colony } from '~data/index';
+import { ValuesType } from '~pages/ExpenditurePage/types';
+
+import ChangeItem from '../../ChangedMultiple/ChangeItem';
+import { ValueOf } from '../../ChangedValues/ChangedValues';
+
```
Comment (jakubcolony) [inline src/modules/dashboard/components/Dialogs/EditExpenditureDialog/ChangedStaged/ChangedMilestone/ChangedMilestone.tsx:11]: Can remove the empty line?

File: src/modules/pages/components/ExpenditurePage/utils.ts (hunk 1 above)
Comment (jakubcolony) [reply src/modules/pages/components/ExpenditurePage/utils.ts:332]: Looks like this one slipped through your fix @agnieszkajarosikloj

**Review by jakubcolony — APPROVED** [review]
Nice work @agnieszkajarosikloj! 

Just noticed there's one typo left

**Comment by agnieszkajarosikloj** [comment]
@willm30 This is ready for re-review.

**Review by willm30 — COMMENTED** [review]
Where it says "Change Staged", should that read "Change Token"? That was the change I made on that line. 

editp

**Comment by agnieszkajarosikloj** [comment]
@willm30 I have changed header. I left one header for amount and user. Can you have a look?


Screenshot 2022-10-17 at 11 56 23

**Comment by willm30** [comment]
@agnieszkajarosikloj Would it be possible to put the "Changed User" on its own line? Also, the user avatar looks to be larger than the token icons. I think we could make these all the same size.

**Comment by agnieszkajarosikloj** [comment]
@willm30 headers and styles have been changed. Can you have a look please?

**Review by willm30 — APPROVED** [review]
Nice work! One final request (sorry for the back and forth): I think we can omit "Staged" from "Change Staged Amount" and "Change Staged User" to leave "Change Amount" and "Change User". Thereby maintaining consistency between headers. 

And also, the "title" and "description" in "Change title" and "Change description" aren't capitalised like in other headers. Could we update that too?

**Pull request merged**